### PR TITLE
Prometheus: Framing performance improvements

### LINKED ISSUE #43369: Increased memory usage when querying Prometheus datasources since 8.3.x
***
_@bohandley update September 12, 2022_
**Description:** Memory usage increased with Prometheus queries
**Acceptance Criteria:** Improve performance of Prometheus query memory usage by successfully implementing the streaming parser.
**Status:** 
@toddtreece introduced the streaming parser to prometheus and began working on bring it to parity with the old prom client. 
https://github.com/grafana/grafana/pull/49858
https://github.com/grafana/grafana/pull/50206
@aocenas helped our squad with a plan to bring the streaming to parity by comparing it with the old client.
https://github.com/grafana/grafana/issues/52738
@ismail is currently assigned the tasks to bring it to parity and remove the old client. 
@toddtreece and @ryantxu have a plan to test the memory usage for Prometheus queries using real world testing as well as testing in staging and ops using conprof/parca (and now pyroscope?). This work is in progress and we are working to align everyone so that we can improve memory usage for Prometheus queries. 
***

**What happened**:

When querying Prometheus datasources the memory usage of Grafana server has increased since Grafana `8.3.x` when compared to `8.2.x`.

Depending on the size of the result set, the memory usage has increased by 1.5x to 3x times, when comparing `8.3.3` to `8.2.7`.

**What you expected to happen**:

Memory usage to not increase, or to not increase as sharply.

**How to reproduce it (as minimally and precisely as possible)**:

1. Launch a Prometheus instance
2. Launch a 8.2.7 Grafana instance (instance A)
3. Launch a 8.3.3 Grafana instance (instance B)
4. Add scrape configs for both Grafana instances to your Prometheus instance
5. Add Prometheus instance as datasource to both Grafana instances
6. Query (e.g. in Explore) any metric (e.g. Go GC duration) on instance A a few times
7. Query (e.g. in Explore) any metric (e.g. Go GC duration) on instance B a few times
8. Query `go_memstats_alloc_bytes` for instance A and instance B
9. You will see higher value for instance B

**Anything else we need to know?**:

The issue has been caused by the fact that Prometheus datasource has been refactored from a frontend datasource to a backend datasource and since 8.3 all queries have to be processed in Grafana server:

* Grafana Frontend sends the request from the browser to the Grafana server
* Grafana server calculates the necessary Prometheus query
* Grafana server sends calculated query to Prometheus API
* Grafana server receives and parses the response
* Grafana server converts the response to DataFrames
* Grafana server sends the DataFrames back to Grafana Frontend

**Environment**:
- Grafana version: 8.3.x
- Data source type & version: Prometheus
- OS Grafana is installed on: Google Container-Optimised OS
- User OS & Browser: MacOS 12.1 / Safari 15.2
- Grafana plugins: Default only
- Others: N/A

### PR DESCRIPTION
**Pull request by toddtreece**
## What this PR does / why we need it

This is an attempt to reduce memory and CPU usage in the prometheus data source backend. See https://github.com/grafana/grafana/issues/43369 for more info

### MatrixToDataFrames

There was a  improvement here by converting from multi-frame series to wide format. There were also a lot of extra allocations caused by type casting `value := float64(k.Value)`.

```
name                   old time/op    new time/op    delta
MatrixToDataFrames-32    7.35ms ± 4%    3.94ms ± 7%  -46.43%  (p=0.000 n=9+10)

name                   old alloc/op   new alloc/op   delta
MatrixToDataFrames-32    6.10MB ± 0%    1.09MB ± 1%  -82.04%  (p=0.000 n=9+10)

name                   old allocs/op  new allocs/op  delta
MatrixToDataFrames-32      181k ± 0%        3k ± 1%  -98.43%  (p=0.000 n=10+10)
```

### VectorToDataFrames
This uses essentially the same code as `MatrixToDataFrames`, but the room for improvement is limited due to the shape of the vector response.
```
name                   old time/op    new time/op    delta
VectorToDataFrames-32    24.2ms ± 9%    16.8ms ± 1%  -30.62%  (p=0.000 n=10+8)

name                   old alloc/op   new alloc/op   delta
VectorToDataFrames-32    16.0MB ± 0%     8.5MB ± 0%  -46.95%  (p=0.000 n=10+9)

name                   old allocs/op  new allocs/op  delta
VectorToDataFrames-32      275k ± 0%      145k ± 0%  -47.35%  (p=0.000 n=10+10)
```

### ScalarToDataFrames

It doesn't seem like there is any room for improvement on this one since the response is one row with two columns.

### ExemplarToDataFrames

I looked at this one but I don't currently understand the expected behavior well enough to make modifications, so I'll address this in a separate PR.

## Which issue(s) this PR fixes

* #43369 

Not sure if "fix" is appropriate since we can't avoid additional resource usage when switching to backend data sources

### TODO

* [x] MatrixToDataFrames
* [ ] ExemplarToDataFrames (see note above)
* [x] ScalarToDataFrames
* [x] VectorToDataFrames

### CODE REVIEW

File: pkg/tsdb/prometheus/time_series_query.go [hunk 1]
```diff
@@ -292,34 +292,61 @@ func interpolateVariables(expr string, interval time.Duration, timeRange time.Du
 	return expr
 }
 
-func matrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data.Frames) data.Frames {
+// MatrixToDataFrames marshals a matrix result to data frames
+func MatrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data.Frames) data.Frames {
```
Comment (toddtreece) [inline pkg/tsdb/prometheus/time_series_query.go:296]: i needed to export this to make it available in the `prometheus_test` package

File: pkg/tsdb/prometheus/time_series_query_test.go [hunk 2]
```diff
@@ -572,7 +572,7 @@ func TestPrometheus_parseTimeSeriesResponse(t *testing.T) {
 		require.NoError(t, err)
 
 		require.Len(t, res, 1)
-		require.Equal(t, res[0].Name, "legend Application")
+		require.Equal(t, res[0].Name, "")
```
Comment (toddtreece) [inline pkg/tsdb/prometheus/time_series_query_test.go:575]: @kylebrandt i'm converting from multi-frame series to wide series in this PR. i was hoping you could take a look and give some advice on the best approach for naming the wide frame.

### main
* each frame name in the multi-frame series is set to the output of `formatLegend`
* value field has `DisplayNameFromDS` set to the output of `formatLegend`
https://github.com/grafana/grafana/blob/43c81ddd23f574ab2bba8c23e008a6cc003028f3/pkg/tsdb/prometheus/time_series_query.go#L313-L319

### this PR
* wide frame name set to `""`
* value field has `DisplayNameFromDS` set to the output of `formatLegend`

https://github.com/grafana/grafana
/blob/4984e4441292cb2bc57ede117a028d8906e0d8ef/pkg/tsdb/prometheus/framer.go#L84-L94

does that seem like the best approach?

Comment (kylebrandt) [reply pkg/tsdb/prometheus/time_series_query_test.go:575]: I honestly don't know. I thought:
 - Value Field Name is metric name
 - Labels on are also on that same field

This is what made sense to me - I never noticed the `data.TimeSeriesValueFieldName` before, that got added without me noticing. So how names get used from the Field Name vs Frame Name across Grafana is not something I know.

Comment (kylebrandt) [reply pkg/tsdb/prometheus/time_series_query_test.go:575]: I would guess it is probably okay to leave the Frame name blank on the wide frame though.

Comment (toddtreece) [reply pkg/tsdb/prometheus/time_series_query_test.go:575]: 👍 thanks. updated in 2a87f97 to use the output of `formatLegend` as the value field name

Comment (toddtreece) [reply pkg/tsdb/prometheus/time_series_query_test.go:575]: this seemed to change the output in grafana when comparing to main, so i reverted in 7a677cd

**Comment by torkelo** [comment]
This would be good to implement behind a feature toggle as this will likely break a lot of dashboards (at least for old panels).

**Comment by bergquist** [comment]
> This would be good to implement behind a feature toggle as this will likely break a lot of dashboards (at least for old panels).

Yes please! This is the bread and butter of Grafana Cloud so we want to enable and disable this without having to rollback Grafana version.

**Comment by toddtreece** [comment]
@torkelo @bergquist i have been talking with @kylebrandt and @ryantxu about this in slack to attempt to make sure that the wide frames are backward compatible. if that's possible (and backed up by tests) do you still think a feature flag would still be useful here?

**Comment by toddtreece** [comment]
i moved this back to draft for now after some discussions with @gabor and @ivanahuckova about balancing backwards compatibility, code readability, and performance/memory usage optimizations

**Comment by torkelo** [comment]
Not sure it's possible to make this fully backward compatible (Factoring in transformations, old panels, etc)

**Comment by torkelo** [comment]
But I do think moving to wide frames is a good idea

**Comment by torkelo** [comment]
But it's a bit sensitve, when all the series will share the same time field, it's important that the same behavior for missing values (nulls) is preserved, and for missing values due to unaligned series it should be undefined. We need to be able to separate missing value caused by no scrape data, from missing value caused by aligning all series into a single frame with shared time field

**Comment by kylebrandt** [comment]
"We need to be able to separate missing value caused by no scrape data, from missing value caused by aligning all series into a single frame with shared time field"

Within a Frame that could only be achieved I think with a new metadata feature (in a generic data source sense). Data sources can set values to things like null and NaN, so without metadata there is no way to distinguish if it was original data or data added to make the time index shared across metrics when it wasn't originally.

So the multi-frame format is better than wide from the standpoint of preservation of original data if there is not a shared time index for all series. I have conversion notes here https://docs.google.com/document/d/1Dd8jsrU7_ys7MqDCDEbUE8NfKmtUJ_MX2_HN4PY8YOk/edit#heading=h.iwgmengn8i6q , many/multi is the only logical time series structure that you can convert to without modifying data.

**Comment by toddtreece** [comment]
closing this one since there are still some things to figure out with wide frames. i'll open a new PR and attempt to migrate any relevant improvements to the current multi-frame output

**Pull request closed without merging**